### PR TITLE
fix(a380x): Fixes for PFD, MFD, GPWS

### LINKED ISSUE #8960: a380x: duplicate names page appears on airway termination
### Aircraft Version

Development

### Build info

```json
master
```

### Describe the bug

Setup flight plan `NZCH/02 GUTBU1P.GRETA NZWN`, revise -> airways on GRETA, enter Y393, then try to enter WARDS as the termination. The duplicate names page appears with a 2nd WARDS that is no on Y393 from GRETA.

### Expected behavior

The FMS should only consider fixes on the airway.

### Steps to reproduce

As in description.

### References (optional)

_No response_

### Additional info (optional)

_No response_

### Discord Username (optional)

_No response_

### LINKED ISSUE #8961: a380x: STD flashing on PFD
### Aircraft Version

Development

### Build info

```json
master
```

### Describe the bug

STD flashes on the PFD at FL220 in FMS DES phase when no TRANS FL is entered on the PERF APPR page.

### Expected behavior

STD should not flash in this scenario.

### Steps to reproduce

Cruise at FL220, or any other higher alt with no TRANS FL set. Enter DES phase, and note STD flashing on the PFD.

### References (optional)

_No response_

### Additional info (optional)

_No response_

### Discord Username (optional)

_No response_

### PR DESCRIPTION
**Pull request by flogross89**
Fixes #8961 
Fixes #8960 

## Summary of Changes

- Make GPWS altitude warning C-chord cancellable through master warning p/b
- Rename FMC files from .tsx to .ts (no JSX)
- Improve FMS FPLN page transition alt logic (print all altitudes as altitude if no trans alt defined)
- Fix RMP memo logic for RMP OFF and xxx ON RMP 3
- Choose fix on airway on FMS/FPLN/AIRWAYS page, don't deduplicate arbitrary fixes
- Fix PFD QNH flashing if no trans alt / lvl set
- Fixes on FUEL&LOAD page: BLOCK field disappears instantly if engines running and page is opened, not only after 1 second

## Additional context

Discord username (if different from GitHub):

## Testing instructions

1. Confirms fixes for #8961 , #8960 
2. In clean config, overshoot FCU alt, confirm that continuous c-chord can be heard and can be cancelled through the left or right master warning push button
3. Enable/disable different RMPs and verify that RMP MEMOs in EWD are plausible (i.e. no duplicates such as RMP 1 OFF and RMP 1+3 OFF)

## How to download the PR for QA

Every new commit to this PR will cause new A32NX and A380X artifacts to be created, built, and uploaded.

1. Make sure you are signed in to GitHub
1. Click on the **Checks** tab on the PR
1. On the left side, find and click on the **PR Build** tab
1. Click on either **flybywire-aircraft-a320-neo** or **flybywire-aircraft-a380-842** download link at the bottom of the page

### CODE REVIEW

File: fbw-a380x/src/systems/instruments/src/MFD/FMC/FmcAircraftInterface.ts [hunk 1]
```diff
@@ -209,34 +210,26 @@ export class FmcAircraftInterface {
     if (!this.flightPlanService.hasActive) {
       return;
     }
-    const originTransitionAltitude = this.flightPlanService.active.performanceData.transitionAltitude; // as altitude
-    const destinationTransitionLevel = this.flightPlanService.active.performanceData.transitionLevel ?? 18_000; // as FL
 
     this.arincTransitionAltitude.setBnrValue(
-      originTransitionAltitude !== null ? originTransitionAltitude : 0,
-      originTransitionAltitude !== null
+      this.flightPlanService.active.performanceData.transitionAltitude ?? 2_500, // as altitude
```
Comment (tracernz) [inline fbw-a380x/src/systems/instruments/src/MFD/FMC/FmcAircraftInterface.ts:215]:
```suggestion
      this.flightPlanService.active.performanceData.transitionAltitude ?? 0, // as altitude
```
Should be 0/NCD when not entered.

File: fbw-a380x/src/systems/instruments/src/MFD/FMC/FmcAircraftInterface.ts [hunk 2]
```diff
@@ -209,34 +210,26 @@ export class FmcAircraftInterface {
...
       0,
     );
 
-    // Delete once new PFD is integrated
-    SimVar.SetSimVarValue('L:AIRLINER_TRANS_ALT', 'Number', originTransitionAltitude ?? 0);
-
     this.arincTransitionLevel.setBnrValue(
-      destinationTransitionLevel !== undefined ? destinationTransitionLevel : 0,
-      destinationTransitionLevel !== undefined
+      this.flightPlanService.active.performanceData.transitionLevel ?? 25, // as FL
```
Comment (tracernz) [inline fbw-a380x/src/systems/instruments/src/MFD/FMC/FmcAircraftInterface.ts:225]:
```suggestion
      this.flightPlanService.active.performanceData.transitionLevel ?? 0, // as FL
```
As above, 0/NCD

File: fbw-a380x/src/systems/instruments/src/MFD/FMC/fmgc.ts [hunk 3]
```diff
@@ -315,6 +315,18 @@ export class FmgcDataService implements Fmgc {
     return (zfw / 1000) * 2204.625;
   }
 
+  /** in kg */
+  public getGrossWeight(): Kilograms {
+    // Value received from FQMS, or falls back to ZFW + FOB
+    const zfw = this.data.zeroFuelWeight.get();
+
+    let fmGW = SimVar.GetSimVarValue('TOTAL WEIGHT', 'kilogram');
```
Comment (tracernz) [inline fbw-a380x/src/systems/instruments/src/MFD/FMC/fmgc.ts:323]: There's no way for the plane to measure it's weight without a ZFW entry + either FoB after engine start, or block entry before engine start. We should not be reading this var.

File: fbw-a380x/src/systems/instruments/src/MFD/pages/FMS/F-PLN/MfdFmsFplnAirways.tsx [hunk 4]
```diff
@@ -251,24 +251,38 @@ class AirwayLine extends DisplayComponent<AirwayLineProps> {
...
+
+                if (fixes.length > 1) {
+                  const dedup = await this.props.fmc.deduplicateFacilities(fixes);
+                  if (dedup !== undefined) {
+                    chosenFix = dedup;
+                  }
                 }
               }
 
+              if (!chosenFix) return false;
```
Comment (tracernz) [inline fbw-a380x/src/systems/instruments/src/MFD/pages/FMS/F-PLN/MfdFmsFplnAirways.tsx:284]:
```suggestion
              if (!chosenFix) {
                return false;
              }
```

File: fbw-a380x/src/systems/instruments/src/MFD/pages/FMS/F-PLN/MfdFmsFplnAirways.tsx [hunk 5]
```diff
@@ -251,24 +251,38 @@ class AirwayLine extends DisplayComponent<AirwayLineProps> {
...
-                const dedup = await this.props.fmc.deduplicateFacilities(fixes);
-                if (dedup !== undefined) {
-                  chosenFix = dedup;
+              let chosenFix: Fix | undefined = undefined;
+
+              if (this.viaField.get() !== 'DCT') {
+                try {
+                  chosenFix = await this.props.pendingAirways.fixAlongTailAirway(v);
+                } catch (msg: unknown) {
+                  if (msg instanceof FmsError) this.props.fmc.showFmsErrorMessage(msg.type);
```
Comment (tracernz) [inline fbw-a380x/src/systems/instruments/src/MFD/pages/FMS/F-PLN/MfdFmsFplnAirways.tsx:266]:
```suggestion
                  if (msg instanceof FmsError) {
                    this.props.fmc.showFmsErrorMessage(msg.type);
                  }
```

File: fbw-a380x/src/systems/instruments/src/MFD/pages/FMS/MfdFmsPerf.tsx [hunk 6]
```diff
@@ -2206,7 +2206,7 @@ export class MfdFmsPerf extends FmsPage<MfdFmsPerfProps> {
                     <InputField<number>
                       dataEntryFormat={
                         new AltitudeOrFlightLevelFormat(
-                          this.transFl.map((it) => (it ?? 180) * 100),
+                          this.transFl.map((it) => (it ?? 25) * 100),
```
Comment (tracernz) [inline fbw-a380x/src/systems/instruments/src/MFD/pages/FMS/MfdFmsPerf.tsx:2209]: Should be no default value for transFl.

File: fbw-a380x/src/systems/instruments/src/MFD/pages/FMS/MfdFmsPerf.tsx [hunk 7]
```diff
@@ -2538,14 +2538,7 @@ export class MfdFmsPerf extends FmsPage<MfdFmsPerfProps> {
...
                         selectedIndex={this.apprSelectedFlapsIndex}
-                        onModified={(v) => {
-                          this.props.fmcService.master?.fmgc.data.approachFlapConfig.set(v + 3);
-                          if (v === 0) {
-                            SimVar.SetSimVarValue('L:A32NX_SPEEDS_LANDING_CONF3', 'boolean', true);
-                          } else {
-                            SimVar.SetSimVarValue('L:A32NX_SPEEDS_LANDING_CONF3', 'boolean', false);
-                          }
-                        }}
+                        onModified={(v) => this.props.fmcService.master?.fmgc.data.approachFlapConfig.set(v + 3)}
```
Comment (tracernz) [inline fbw-a380x/src/systems/instruments/src/MFD/pages/FMS/MfdFmsPerf.tsx:2541]:
```suggestion
                        onModified={(v) => this.props.fmcService.master?.fmgc.data.approachFlapConfig.set(v + 3)}
```
This kind of thing is a bug waiting to happen IMO. There should be an explicit mapping from the UI value to the enum approachFlagConfig expects.

File: fbw-a380x/src/systems/instruments/src/PFD/AltitudeIndicator.tsx [hunk 8]
```diff
@@ -781,18 +782,16 @@ class AltimeterIndicator extends DisplayComponent<AltimeterIndicatorProps> {
   private handleBlink() {
     if (this.mode.get() === 'STD') {
       if (
-        this.flightPhase > 3 &&
-        this.transLvlAr.isNormalOperation() &&
-        100 * this.transLvlAr.value > this.props.altitude.get()
+        this.fmgcFlightPhase > FmgcFlightPhase.Cruise &&
+        100 * (this.transLvlAr.isNormalOperation() ? this.transLvlAr.value : 25) > this.props.altitude.get()
```
Comment (tracernz) [inline fbw-a380x/src/systems/instruments/src/PFD/AltitudeIndicator.tsx:786]: No default for trans level. It should not blink if the trans level is NCD or FW.

File: fbw-a380x/src/systems/instruments/src/PFD/AltitudeIndicator.tsx [hunk 9]
```diff
@@ -781,18 +782,16 @@ class AltimeterIndicator extends DisplayComponent<AltimeterIndicatorProps> {
...
         this.stdGroup.instance.classList.add('BlinkInfinite');
       } else {
         this.stdGroup.instance.classList.remove('BlinkInfinite');
       }
     } else if (
-      this.flightPhase <= 3 &&
-      this.transAltAr.isNormalOperation() &&
-      this.transAltAr.value < this.props.altitude.get()
+      this.fmgcFlightPhase <= FmgcFlightPhase.Cruise &&
+      (this.transAltAr.isNormalOperation() ? this.transAltAr.value : 2_500) < this.props.altitude.get()
```
Comment (tracernz) [inline fbw-a380x/src/systems/instruments/src/PFD/AltitudeIndicator.tsx:794]: No default for trans alt. It should not blink if the trans level is NCD or FW.

File: fbw-a380x/src/systems/systems-host/systems/LegacyGpws.ts [hunk 10]
```diff
@@ -4,888 +4,935 @@ import { LegacySoundManager, soundList } from 'systems-host/systems/LegacySoundM
 import { A380X_DEFAULT_RADIO_AUTO_CALL_OUTS, A380XRadioAutoCallOutFlags } from '../../shared/src/AutoCallOuts';
```
Comment (tracernz) [inline fbw-a380x/src/systems/systems-host/systems/LegacyGpws.ts:4]: Please split the format changes into a separate commit. It's impossible to review with them mixed in.

**Review by tracernz — CHANGES_REQUESTED** [review]

File: fbw-a380x/src/systems/instruments/src/MFD/pages/FMS/MfdFmsPerf.tsx (hunk 6 above)
Comment (flogross89) [reply fbw-a380x/src/systems/instruments/src/MFD/pages/FMS/MfdFmsPerf.tsx:2209]: fixed, trans alt/lvl are now nullable

File: fbw-a380x/src/systems/instruments/src/PFD/AltitudeIndicator.tsx (hunk 8 above)
Comment (flogross89) [reply fbw-a380x/src/systems/instruments/src/PFD/AltitudeIndicator.tsx:786]: roger, reverting to the previous logic

File: fbw-a380x/src/systems/systems-host/systems/LegacyGpws.ts (hunk 10 above)
Comment (flogross89) [reply fbw-a380x/src/systems/systems-host/systems/LegacyGpws.ts:4]: That's why https://github.com/flybywiresim/aircraft/pull/9013 exists and has to be merged before this one

File: fbw-a380x/src/systems/instruments/src/PFD/AltitudeIndicator.tsx (hunk 9 above)
Comment (flogross89) [reply fbw-a380x/src/systems/instruments/src/PFD/AltitudeIndicator.tsx:794]: same here as before, will use old logic

File: fbw-a380x/src/systems/instruments/src/MFD/FMC/fmgc.ts (hunk 3 above)
Comment (flogross89) [reply fbw-a380x/src/systems/instruments/src/MFD/FMC/fmgc.ts]: took that logic now

File: fbw-a380x/src/systems/systems-host/systems/LegacyGpws.ts (hunk 10 above)
Comment (tracernz) [reply fbw-a380x/src/systems/systems-host/systems/LegacyGpws.ts:4]: But the changes seem to be mixed in the same commit here?

Comment (flogross89) [reply fbw-a380x/src/systems/systems-host/systems/LegacyGpws.ts]: Just had a look again, I didn't change anything in LegacyGpws, don't know why I linted it in the first place then :/ I reverted the change on this file.

File: fbw-a380x/src/systems/instruments/src/MFD/FMC/FmcAircraftInterface.ts [hunk 11]
```diff
@@ -209,34 +210,26 @@ export class FmcAircraftInterface {
...
       return;
     }
-    const originTransitionAltitude = this.flightPlanService.active.performanceData.transitionAltitude; // as altitude
-    const destinationTransitionLevel = this.flightPlanService.active.performanceData.transitionLevel ?? 18_000; // as FL
 
     this.arincTransitionAltitude.setBnrValue(
-      originTransitionAltitude !== null ? originTransitionAltitude : 0,
-      originTransitionAltitude !== null
+      this.flightPlanService.active.performanceData.transitionAltitude ?? 0, // as altitude
+      this.flightPlanService.active.performanceData.transitionAltitude != null
```
Comment (tracernz) [inline fbw-a380x/src/systems/instruments/src/MFD/FMC/FmcAircraftInterface.ts:216]:
```suggestion
      this.flightPlanService.active.performanceData.transitionAltitude !== null
```

File: fbw-a380x/src/systems/instruments/src/MFD/FMC/FmcAircraftInterface.ts [hunk 12]
```diff
@@ -209,34 +210,26 @@ export class FmcAircraftInterface {
...
     );
 
-    // Delete once new PFD is integrated
-    SimVar.SetSimVarValue('L:AIRLINER_TRANS_ALT', 'Number', originTransitionAltitude ?? 0);
-
     this.arincTransitionLevel.setBnrValue(
-      destinationTransitionLevel !== undefined ? destinationTransitionLevel : 0,
-      destinationTransitionLevel !== undefined
+      this.flightPlanService.active.performanceData.transitionLevel ?? 0, // as FL
+      this.flightPlanService.active.performanceData.transitionLevel != null
```
Comment (tracernz) [inline fbw-a380x/src/systems/instruments/src/MFD/FMC/FmcAircraftInterface.ts:226]:
```suggestion
      this.flightPlanService.active.performanceData.transitionLevel !== null
```

File: fbw-a380x/src/systems/instruments/src/MFD/pages/FMS/F-PLN/MfdFmsFpln.tsx [hunk 13]
```diff
@@ -309,10 +309,9 @@ export class MfdFmsFpln extends FmsPage<MfdFmsFplnProps> {
       }
 
       if (leg instanceof FlightPlanLeg) {
-        const transAlt = this.loadedFlightPlan.performanceData.transitionAltitude ?? 18_000;
-        const transLevelAsAlt = Number.isFinite(this.loadedFlightPlan.performanceData.transitionLevel)
-          ? (this.loadedFlightPlan.performanceData.transitionLevel ?? 18_000) * 100
-          : 18_000;
+        const transAlt = this.loadedFlightPlan.performanceData.transitionAltitude;
+        const transLevel = this.loadedFlightPlan.performanceData.transitionLevel;
+        const transLevelAsAlt = transLevel != null ? transLevel * 100 : null;
```
Comment (tracernz) [inline fbw-a380x/src/systems/instruments/src/MFD/pages/FMS/F-PLN/MfdFmsFpln.tsx:314]:
```suggestion
        const transLevelAsAlt = transLevel !== null ? transLevel * 100 : null;
```

File: fbw-a380x/src/systems/instruments/src/MFD/pages/FMS/F-PLN/MfdFmsFpln.tsx [hunk 14]
```diff
@@ -1389,7 +1388,8 @@ class FplnLegLine extends DisplayComponent<FplnLegLineProps> {
     let altStr: VNode = <span>-----</span>;
     const previousRow = this.props.previousRow.get();
     if (data.altitudePrediction) {
-      const isBelowTransAlt = data.altitudePrediction < (data.transitionAltitude ?? 18_000);
+      const isBelowTransAlt =
+        data.transitionAltitude != null ? data.altitudePrediction < data.transitionAltitude : true;
```
Comment (tracernz) [inline fbw-a380x/src/systems/instruments/src/MFD/pages/FMS/F-PLN/MfdFmsFpln.tsx:1392]:
```suggestion
        data.transitionAltitude !== null ? data.altitudePrediction < data.transitionAltitude : true;
```

File: fbw-a380x/src/systems/instruments/src/MFD/pages/FMS/F-PLN/MfdFmsFpln.tsx [hunk 15]
```diff
@@ -1425,7 +1425,8 @@ class FplnLegLine extends DisplayComponent<FplnLegLineProps> {
       );
     }
     if (data.hasAltitudeConstraint && data.altitudeConstraint?.altitude1 && !data.altitudePrediction) {
-      const isBelowTransAlt = data.altitudeConstraint.altitude1 < (data.transitionAltitude ?? 18_000);
+      const isBelowTransAlt =
+        data.transitionAltitude != null ? data.altitudeConstraint.altitude1 < data.transitionAltitude : true;
```
Comment (tracernz) [inline fbw-a380x/src/systems/instruments/src/MFD/pages/FMS/F-PLN/MfdFmsFpln.tsx:1429]:
```suggestion
        data.transitionAltitude !== null ? data.altitudeConstraint.altitude1 < data.transitionAltitude : true;
```

File: fbw-a380x/src/systems/instruments/src/MFD/pages/FMS/MfdFmsPerf.tsx [hunk 16]
```diff
@@ -2206,7 +2207,7 @@ export class MfdFmsPerf extends FmsPage<MfdFmsPerfProps> {
                     <InputField<number>
                       dataEntryFormat={
                         new AltitudeOrFlightLevelFormat(
-                          this.transFl.map((it) => (it ?? 180) * 100),
+                          this.transFl.map((it) => (it != null ? it * 100 : null)),
```
Comment (tracernz) [inline fbw-a380x/src/systems/instruments/src/MFD/pages/FMS/MfdFmsPerf.tsx:2210]:
```suggestion
                          this.transFl.map((it) => (it !== null ? it * 100 : null)),
```

File: fbw-a380x/src/systems/instruments/src/MFD/pages/common/DataEntryFormats.tsx [hunk 17]
```diff
@@ -109,29 +109,31 @@ export class AltitudeOrFlightLevelFormat implements DataEntryFormat<number> {
...
     maxValue: Subscribable<number> = Subject.create(maxCertifiedAlt),
   ) {
     minValue.sub((val) => (this.minValue = val), true);
     maxValue.sub((val) => (this.maxValue = val), true);
 
-    transAlt.sub((val) => {
-      this.transAlt = val;
-      this.reFormatTrigger.notify();
-    });
+    if (transAlt != null) {
```
Comment (tracernz) [inline fbw-a380x/src/systems/instruments/src/MFD/pages/common/DataEntryFormats.tsx:124]:
```suggestion
    if (transAlt !== null) {
```

File: fbw-a380x/src/systems/instruments/src/MFD/pages/common/DataEntryFormats.tsx [hunk 18]
```diff
@@ -109,29 +109,31 @@ export class AltitudeOrFlightLevelFormat implements DataEntryFormat<number> {
...
+      });
+    }
   }
 
   public format(value: number) {
     if (value === null || value === undefined) {
       return [this.placeholder, null, 'FT'] as FieldFormatTuple;
     }
-    if (value >= this.transAlt) {
+    if (this.transAlt != null && value >= this.transAlt) {
```
Comment (tracernz) [inline fbw-a380x/src/systems/instruments/src/MFD/pages/common/DataEntryFormats.tsx:136]:
```suggestion
    if (this.transAlt !== null && value >= this.transAlt) {
```

**Review by tracernz — APPROVED** [review]
Small fixes with JS being JS, otherwise LGTM.

We should have the typescript-eslint rule for this one but we must not.
-e- tried enabling it but it's a real pain. We have 3 other instances of this already in the repo :sadcat:.

**Comment by pilotseyea350** [comment]
Quality Assurance Tester Report

Discord Username  : PilotEyesA350 
Object of testing : #9014 
Aircraft          : A380X 
Tier of Testing   : 2
Date              : 07/10/2024 

Testing Process: 
1. Load aircraft as usual
2. Turn RMPs off 1 by 1 and observe any inconsistencies on the E/WD memo
3. Fly from LFPG to LFPG
4. Climb to FL220, turn autopilot off and overshoot the altitude selected
5. Once c-chord is heard, press CPT M/W PB, confirm c-chord is turned off
6. Delete TRANS FL in the FMS APPR PERF page
7. Observe STD on PFD is not flashing
8. Park at any stand, reprogram the FMS from LFPG to LFPG
9. Enter RWY26R and any EVX SID
10. From EVX enter Direct EKRAS
11. From EKRAS enter airway UN585 to INGOR
12. Observe INGOR isn't duplicated on the MFD F-PLN page

Testing Results: 
Passed

Negatives:       
N/A

Conclusions: 
floridude you absolute legend

Media:
N/A

**Comment by flogross89** [comment]
> Quality Assurance Tester Report
> 
> Discord Username : PilotEyesA350 Object of testing : #9014 Aircraft : A380X Tier of Testing : 2 Date : 07/10/2024
> 
> Testing Process:
> 
> 1. Load aircraft as usual
> 2. Turn RMPs off 1 by 1 and observe any inconsistencies on the E/WD memo
> 3. Fly from LFPG to LFPG
> 4. Climb to FL220, turn autopilot off and overshoot the altitude selected
> 5. Once c-chord is heard, press CPT M/W PB, confirm c-chord is turned off
> 6. Delete TRANS FL in the FMS APPR PERF page
> 7. Observe STD on PFD is not flashing
> 8. Park at any stand, reprogram the FMS from LFPG to LFPG
> 9. Enter RWY26R and any EVX SID
> 10. From EVX enter Direct EKRAS
> 11. From EKRAS enter airway UN585 to INGOR
> 12. Observe INGOR isn't duplicated on the MFD F-PLN page
> 
> Testing Results: Passed
> 
> Negatives: N/A
> 
> Conclusions: floridude you absolute legend
> 
> Media: N/A

haha love the conclusion, thanks a lot for the test!

**Pull request merged**